### PR TITLE
Feature/refactor player staging

### PR DESCRIPTION
**Pull request by fortypercenttitanium**
#### 0. Make sure your issue is linked:

Needs to be done before #333 

#### Before:
Players connected and were add to a "staging area" array, which would be cleared at the beginning of each round.

#### After:
Players connected are given `isConnected` and `isPlaying` on their player object in the Host Context. These are used to determine dealing of cards, picking of czar, etc... and will be further used to implement reconnects.

#### 2. Implementation:

I removed the staging array and all associated tests and replaced each function that utilized it with an implementation of an object property on the player object. 

#### 3. Possible Issues:

There were a lot of tests that were edited to make this work. I added a few test cases, but I was mostly focused on fixing the tests that were already in place. We may still be missing some unit test edge cases.

#### 4. How to Test:

* Pull branch
* Review code
* Run the game and play, ensuring there are no errors in the consoles
* Try to force edge cases

#### 5. Cleanup
- [x] I've added all TODOs needed
- [x] I've removed all unneeded comments
- [x] I've updated the snapshot tests and checked to make sure they're accurate if they changed
- [x] I've removed all unnecessary `console.log`s

### CODE REVIEW

**Review by I3uckwheat — CHANGES_REQUESTED** [review]
I haven't had a chance to pull and run this, but most of this looks pretty good outside of the `isPlaying` issue that I mentioned.

**Review by wise-king-sullyman — CHANGES_REQUESTED** [review]
Awesome work, I know how much of a pain this must've been!

Apologies for some of the line locations where I requested additional tests being less than helpful, unfortunately since GitHub doesn't let me add comments to lines too far away from changes I couldn't always put them where they would make the most sense.

File: frontend/src/contexts/HostContext/HostReducer.spec.js [hunk 1]
```diff
@@ -586,27 +605,31 @@ describe('reducer', () => {
               { pack: 0, text: 'test' },
```
[the remark this answers is not in the document]
Comment (fortypercenttitanium) [reply frontend/src/contexts/HostContext/HostReducer.spec.js:605]: > I'd also like to see a test added to ensure that the `players` object still includes non-active players.

See test suite 'PLAYER_DISCONNECTED' , first test. If you're looking for something different, let me know!

File: frontend/src/screens/HostPregameScreen/HostPregameScreen.js [hunk 2]
```diff
@@ -132,14 +132,14 @@ ErrorHandler.propTypes = errorHandlerPropTypes;
 
 function LeftPanel() {
   const { state, dispatch } = useContext(HostContext);
-  const { lobbyID, newPlayerStaging } = state;
+  const { lobbyID, playerIDs } = state;
```
[the remark this answers is not in the document]
Comment (fortypercenttitanium) [reply frontend/src/screens/HostPregameScreen/HostPregameScreen.js:135]: > We need to enforce that the game can't be started with only disconnected players, I just tested and this causes a crash. Suggest replacing `playerIDs` with only players with `status: 'staging'` and also adding a test to ensure this requirement is met/functioning.

Very nice catch.

**Review by wise-king-sullyman — DISMISSED** [review]

File: frontend/src/contexts/HostContext/HostReducer.js [hunk 3]
```diff
@@ -163,21 +161,29 @@ function removeSubmittedCards(state) {
...
     },
   };
 
   if (
     removingCzar &&
-    state.playerIDs.filter((id) => state.players[id].isConnected).length > 1 &&
+    state.playerIDs.filter(
+      (id) =>
+        state.players[id].status === 'playing' ||
+        state.players[id].status === 'staging',
```
Comment (I3uckwheat) [inline frontend/src/contexts/HostContext/HostReducer.js:185]: We could use a `.includes` here to clean this up:

`['playing', 'staging'].includes(state.players[id].status`

File: frontend/src/contexts/HostContext/HostReducer.js [hunk 4]
```diff
@@ -212,7 +218,11 @@ function kickPlayer(state, { playerId }) {
 
   if (
     removingCzar &&
-    newPlayerIds.filter((id) => state.players[id].isConnected).length > 1 &&
+    newPlayerIds.filter(
+      (id) =>
+        state.players[id].status === 'playing' ||
+        state.players[id].status === 'staging',
```
Comment (I3uckwheat) [inline frontend/src/contexts/HostContext/HostReducer.js:224]: We could use a .includes here to clean this up:

`['playing', 'staging'].includes(state.players[id].status`

**Review by I3uckwheat — CHANGES_REQUESTED** [review]
2 smallllll things, otherwise, great work on this large project!

**Review by I3uckwheat — APPROVED** [review]

**Review by wise-king-sullyman — APPROVED** [review]
